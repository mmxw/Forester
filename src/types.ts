/**
 * types that are shared between several modules
 */
import type {StackScreenProps} from '@react-navigation/stack';
import type {Contact} from 'react-native-select-contact';

/**
 * Users can create a plant for one of their contacts,
 * and the app will indicate if they need to "water" the plant
 * (stay in touch with the friend or family member or colleague)
 */
export type Plant = Readonly<{
  plantId: PlantId;
  contact: Contact;
  lastWatered: Date;
  waterFrequency: WaterFrequency;
  speciesId: SpeciesId;
}>;

/**
 * Plants get droopy and then depart if they aren't
 * watered frequently enough.
 */
export type PlantState = 'happy' | 'droopy' | 'departed';

/**
 * A plant has a species, such as "hosue plant" or "cactus",
 * which determines how it displays.
 */
export type Species = Readonly<{
  readonly id: SpeciesId;
  readonly name: string;
  readonly appearances: Record<PlantState, PlantAppearance>;
}>;

/**
 * Represents in a Plant in a way that is
 * easier to use in the UI.
 */
export type UIPlant = Readonly<{
  name: string;
  appearance: PlantAppearance;
  lastWatered: Date;
  waterFrequency: WaterFrequency;
  nextWatering: Date;
  state: PlantState;
}>;

export type PlantId = string & {readonly plantIdBrand: unique symbol};
export type SpeciesId = string & {readonly speciesId: unique symbol};

export type PlantAppearance = Readonly<{
  kind: 'emoji';
  emoji: string;
}>;

export type WaterFrequency = Readonly<{
  unit: 'days' | 'weeks' | 'months';
  number: number;
}>;

type RootStackParamList = Readonly<{
  Home: undefined;
  PlantChoice: {
    contact: Contact;
  };
}>;

/**
 * Enables screens to have well-typed props
 */
export type ScreenProp<
  Screen extends keyof RootStackParamList
> = StackScreenProps<RootStackParamList, Screen>;
