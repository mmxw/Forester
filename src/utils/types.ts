/**
 * types that are shared between several modules
 */
import type {StackScreenProps} from '@react-navigation/stack';
import type {Contact} from 'react-native-select-contact';

/**
 * Users can create a plant for one of their contacts,
 * and the app will indicate if they need to "water" the plant
 * (stay in touch with the friend or family member or colleague)
 */
export interface Plant {
  readonly plantId: PlantId;
  readonly contact: Contact;
  readonly lastWatered: Date;
  readonly waterFrequency: WaterFrequency;
  readonly speciesId: SpeciesId;
}

/**
 * Plants get droopy and then depart if they aren't
 * watered frequently enough.
 */
export type PlantState = 'happy' | 'droopy' | 'departed';

/**
 * A plant has a species, such as "hosue plant" or "cactus",
 * which determines how it displays.
 */
export interface Species {
  readonly id: SpeciesId;
  readonly name: string;
  readonly appearances: Record<PlantState, PlantAppearance>;
}

/**
 * Represents in a Plant in a way that is
 * easier to use in the UI.
 */
export interface UIPlant {
  readonly plantId: PlantId;
  readonly name: string;
  readonly appearance: PlantAppearance;
  readonly lastWatered: Date;
  readonly waterFrequency: WaterFrequency;
  readonly nextWatering: Date;
  readonly state: PlantState;
}

export type PlantId = string & {readonly plantIdBrand: unique symbol};
export type SpeciesId = string & {readonly speciesId: unique symbol};

export interface PlantAppearance {
  readonly kind: 'emoji';
  readonly emoji: string;
}

export interface WaterFrequency {
  readonly unit: 'days' | 'weeks' | 'months';
  readonly number: number;
}

type RootStackParamList = Readonly<{
  Home: undefined;
  PlantChoice: {
    contact: Contact;
  };
}>;

/**
 * Enables screens to have well-typed props
 */
export type ScreenProp<
  Screen extends keyof RootStackParamList
> = StackScreenProps<RootStackParamList, Screen>;
