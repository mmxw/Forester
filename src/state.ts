import {atom, useRecoilState, useRecoilValue} from 'recoil';
import type {
  Plant,
  PlantAppearance,
  PlantId,
  PlantKind,
  PlantKindId,
  PlantState,
  UIPlant,
  WaterFrequency,
} from './types';
// needed for uuid
import 'react-native-get-random-values';
import {v4} from 'uuid';
import {Contact} from 'react-native-select-contact';
import {plantToUIPlant} from './plant-calculations';

const plantsState = atom({
  key: 'plantsState',
  default: [] as Plant[],
});

const plantKindsState = atom({
  key: 'plantKindsState',
  default: (() => {
    const id = (s: string) => s as PlantKindId;
    function emojiToAppearances(
      emoji: string,
    ): Record<PlantState, PlantAppearance> {
      const appearance: PlantAppearance = {
        kind: 'emoji',
        emoji,
      };
      return {
        happy: appearance,
        droopy: appearance,
        departed: appearance,
      };
    }

    const plantKinds: PlantKind[] = [
      {
        id: id('1'),
        name: 'Cactus',
        appearances: emojiToAppearances('🌵'),
      },
      {
        id: id('2'),
        name: 'Circle Tree',
        appearances: emojiToAppearances('🌳'),
      },

      {
        id: id('3'),
        name: 'Triangle Tree',
        appearances: emojiToAppearances('🌲'),
      },

      {
        id: id('4'),
        name: 'Clover',
        appearances: emojiToAppearances('🍀'),
      },

      {
        id: id('5'),
        name: 'Palm Tree',
        appearances: emojiToAppearances('🌴'),
      },

      {
        id: id('6'),
        name: 'Tiny Plant',
        appearances: emojiToAppearances('🌱'),
      },

      {
        id: id('7'),
        name: 'Sunflower',
        appearances: emojiToAppearances('🌻'),
      },

      {
        id: id('8'),
        name: 'Pink Blossum',
        appearances: emojiToAppearances('🌸 '),
      },

      {
        id: id('9'),
        name: 'Blossom',
        appearances: emojiToAppearances('🌼 '),
      },
    ];

    if (new Set(plantKinds.map((k) => k.id)).size !== plantKinds.length) {
      throw Error('plant kind ids are not unique');
    }
    return plantKinds;
  })(),
});

export function useAddPlant() {
  const [plants, setPlants] = useRecoilState(plantsState);

  // TODO: prevent more than one plant for the same contact?

  function addPlant({
    plantKindId,
    waterFrequency,
    contact,
  }: {
    plantKindId: PlantKindId;
    waterFrequency: WaterFrequency;
    contact: Contact;
  }) {
    const plant = makePlant({
      contact,
      plantKindId,
      waterFrequency,
    });
    setPlants([...plants, plant]);
  }

  return addPlant;
}

export function usePlantKinds(): PlantKind[] {
  const plantKinds = useRecoilValue(plantKindsState);
  return plantKinds.slice();
}

export function useUIPlants(): UIPlant[] {
  const plants = useRecoilValue(plantsState);
  const plantKinds = useRecoilValue(plantKindsState);
  const now = new Date();

  return plants.map((plant) => plantToUIPlant(plant, plantKinds, now));
}

function makePlant({
  plantKindId,
  waterFrequency,
  contact,
}: {
  plantKindId: PlantKindId;
  waterFrequency: WaterFrequency;
  contact: Contact;
}): Plant {
  return {
    plantId: v4() as PlantId,
    contact,
    plantKindId,
    lastWatered: new Date(),
    waterFrequency,
  };
}
