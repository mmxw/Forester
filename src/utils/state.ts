import {atom, useRecoilState, useRecoilValue} from 'recoil';
import type {
  Plant,
  PlantAppearance,
  PlantId,
  Species,
  SpeciesId,
  PlantState,
  UIPlant,
  WaterFrequency,
} from './types';
import {Contact} from 'react-native-select-contact';
import {plantToUIPlant} from './plant-calculations';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useState, useEffect} from 'react';

let i = 0;
/* istanbul ignore next */
const uuid =
  process.env.NODE_ENV === 'test'
    ? () => '' + ++i
    : (() => {
        // needed for uuid. Blows up when run in tests
        require('react-native-get-random-values');
        return require('uuid').v4;
      })();

const plantsState = atom({
  key: 'plantsState',
  default: [] as Plant[],
});

const speciesArrState = atom({
  key: 'speciesArrState',
  default: (() => {
    const id = (s: string) => s as SpeciesId;
    function emojiToAppearances(
      emoji: string,
    ): Record<PlantState, PlantAppearance> {
      const appearance: PlantAppearance = {
        kind: 'emoji',
        emoji,
      };
      return {
        happy: appearance,
        droopy: appearance,
        departed: appearance,
      };
    }

    const speciesArr: Species[] = [
      {
        id: id('1'),
        name: 'Cactus',
        appearances: emojiToAppearances('🌵'),
      },
      {
        id: id('2'),
        name: 'Circle Tree',
        appearances: emojiToAppearances('🌳'),
      },

      {
        id: id('3'),
        name: 'Triangle Tree',
        appearances: emojiToAppearances('🌲'),
      },

      {
        id: id('4'),
        name: 'Clover',
        appearances: emojiToAppearances('🍀'),
      },

      {
        id: id('5'),
        name: 'Palm Tree',
        appearances: emojiToAppearances('🌴'),
      },

      {
        id: id('6'),
        name: 'Tiny Plant',
        appearances: emojiToAppearances('🌱'),
      },

      {
        id: id('7'),
        name: 'Sunflower',
        appearances: emojiToAppearances('🌻'),
      },

      {
        id: id('8'),
        name: 'Pink Blossum',
        appearances: emojiToAppearances('🌸 '),
      },

      {
        id: id('9'),
        name: 'Blossom',
        appearances: emojiToAppearances('🌼 '),
      },
    ];

    /* istanbul ignore next */
    if (new Set(speciesArr.map((s) => s.id)).size !== speciesArr.length) {
      throw Error('plant species ids are not unique');
    }
    return speciesArr;
  })(),
});

async function storePlantsToAsyncStorage(plants: Plant[]): Promise<void> {
  try {
    await AsyncStorage.setItem('plantsState', JSON.stringify(plants));
  } catch (error) {
    /* istanbul ignore next skip error handling*/
    console.error('cannot store the data in local storage', error.stack);
  }
}

export function useAddPlant() {
  const [plants, setPlants] = useRecoilState(plantsState);

  // TODO: prevent more than one plant for the same contact?
  // https://github.com/mmxw/Forester/issues/11

  async function addPlant({
    speciesId,
    waterFrequency,
    contact,
  }: {
    speciesId: SpeciesId;
    waterFrequency: WaterFrequency;
    contact: Contact;
  }) {
    const plant = makePlant({
      contact,
      speciesId,
      waterFrequency,
    });
    const newPlants = [...plants, plant];
    setPlants(newPlants);
    storePlantsToAsyncStorage(newPlants);
  }

  return addPlant;
}

export function useSpeciesArr(): Species[] {
  const speciesArr = useRecoilValue(speciesArrState);
  return speciesArr.slice();
}

type StoragePlant = Omit<Plant, 'lastWatered'> & {lastWatered: string};

async function getPlantsFromAsyncStorage(): Promise<Plant[]> {
  const jsonString = await AsyncStorage.getItem('plantsState');
  const storagePlants: StoragePlant[] =
    jsonString != null ? JSON.parse(jsonString) : [];
  return storagePlants.map((storagePlant) => ({
    ...storagePlant,
    lastWatered: new Date(storagePlant.lastWatered),
  }));
}

export function useUIPlants(): UIPlant[] | 'loading' {
  const [isLoading, setIsLoading] = useState(true);
  const [plants, setPlants] = useRecoilState(plantsState);
  const speciesArr = useRecoilValue(speciesArrState);
  // using Date.now for easier mocking in tests
  const now = new Date(Date.now());

  useEffect(() => {
    getPlantsFromAsyncStorage()
      .then((storedPlants) => {
        setPlants(storedPlants);
      })
      .catch(
        /* istanbul ignore next skip error handling */ (error) => {
          console.error(`error getting plants from storage ${error.stack}`);
          setPlants([]);
        },
      )
      .finally(() => setIsLoading(false));
  }, [setPlants, setIsLoading]);

  if (isLoading) {
    return 'loading';
  }

  return plants.map((plant) => plantToUIPlant(plant, speciesArr, now));
}

function makePlant({
  speciesId,
  waterFrequency,
  contact,
}: {
  speciesId: SpeciesId;
  waterFrequency: WaterFrequency;
  contact: Contact;
}): Plant {
  return {
    plantId: uuid() as PlantId,
    contact,
    speciesId: speciesId,
    // using Date.now for easier mocking
    lastWatered: new Date(Date.now()),
    waterFrequency,
  };
}
