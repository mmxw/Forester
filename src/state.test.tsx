import React, {useState} from 'react';
import {Text} from 'react-native';
import {RecoilRoot} from 'recoil';
import {useAddPlant, usePlantKinds, useUIPlants} from './state';
import {PlantKind, PlantKindId, UIPlant} from './types';
import {fireEvent, render, waitFor} from '@testing-library/react-native';

/*
Lots of boilerplate so we can get at the recoil state.
*/
function StateTester(): JSX.Element {
  /* testID increments each time we `onPress` */
  const [testIDNum, setTestIdNum] = useState(1);
  const uiPlants = useUIPlants();
  const plantKinds = usePlantKinds();
  const addPlant = useAddPlant();
  function onPress() {
    setTestIdNum(testIDNum + 1);
    addPlant(plantData);
  }
  const plantData = {
    plantKindId: '1' as PlantKindId,
    contact: {
      emails: [],
      name: 'p1',
      phones: [],
      postalAddresses: [],
      recordId: 'r1',
    },
    waterFrequency: {
      number: 1,
      unit: 'days' as const,
    },
  };

  const data = {
    uiPlants,
    plantKinds,
  };

  return (
    <Text testID={'' + testIDNum} onPress={onPress}>
      {JSON.stringify(data, null, 2)}
    </Text>
  );
}

const StateTesterContainer = () => (
  <RecoilRoot>
    <StateTester />
  </RecoilRoot>
);

test('can add a plant', async () => {
  jest
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2020-12-27T17:01:19.102Z').valueOf());

  const rendered = render(<StateTesterContainer />);
  const jsonNode = await rendered.findByTestId('1');

  function getData(): {uiPlants: UIPlant[]; plantKinds: PlantKind[]} {
    const str = jsonNode.children.toString();
    return JSON.parse(str);
  }

  expect(getData().uiPlants).toEqual([]);
  expect(getData().plantKinds).toMatchSnapshot('initial plant kinds');

  fireEvent(jsonNode, 'onPress');

  await waitFor(async () => rendered.findByTestId('2'));
  expect(getData().uiPlants).toMatchSnapshot('UI plants after one added');
});
