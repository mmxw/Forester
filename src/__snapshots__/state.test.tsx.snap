// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add a plant: UI plants after one added 1`] = `
Array [
  Object {
    "appearance": Object {
      "emoji": "🌵",
      "kind": "emoji",
    },
    "lastWatered": "2020-12-27T17:01:19.102Z",
    "name": "p1",
    "nextWatering": "2020-12-27T17:01:19.102Z",
    "state": "happy",
    "waterFrequency": Object {
      "number": 1,
      "unit": "days",
    },
  },
]
`;

exports[`can add a plant: initial plant kinds 1`] = `
Array [
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌵",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌵",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌵",
        "kind": "emoji",
      },
    },
    "id": "1",
    "name": "Cactus",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌳",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌳",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌳",
        "kind": "emoji",
      },
    },
    "id": "2",
    "name": "Circle Tree",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌲",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌲",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌲",
        "kind": "emoji",
      },
    },
    "id": "3",
    "name": "Triangle Tree",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🍀",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🍀",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🍀",
        "kind": "emoji",
      },
    },
    "id": "4",
    "name": "Clover",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌴",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌴",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌴",
        "kind": "emoji",
      },
    },
    "id": "5",
    "name": "Palm Tree",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌱",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌱",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌱",
        "kind": "emoji",
      },
    },
    "id": "6",
    "name": "Tiny Plant",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌻",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌻",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌻",
        "kind": "emoji",
      },
    },
    "id": "7",
    "name": "Sunflower",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌸 ",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌸 ",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌸 ",
        "kind": "emoji",
      },
    },
    "id": "8",
    "name": "Pink Blossum",
  },
  Object {
    "appearances": Object {
      "departed": Object {
        "emoji": "🌼 ",
        "kind": "emoji",
      },
      "droopy": Object {
        "emoji": "🌼 ",
        "kind": "emoji",
      },
      "happy": Object {
        "emoji": "🌼 ",
        "kind": "emoji",
      },
    },
    "id": "9",
    "name": "Blossom",
  },
]
`;
